episodes/12-fire-forget-tasks.md

```diff
@@ -0,0 +1,392 @@
+time (in this case the screen); the system decides in which order the tasks could write to the screen.
+
+::::::::::::::::::::::::::::::::::::::: discussion
+
+## Discussion
+
+- What would happen if in the last code we declare `c` in the main thread?
+- What would happen if we try to modify the value of `x` inside a begin statement?
+
+Discuss your observations.
+
+:::::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::: callout
+
+All variables have a **_scope_** in which they can be used. The variables declared inside a concurrent tasks
+are accessible only by the task. The variables declared in the main task can be read everywhere, but Chapel
+won't allow two concurrent tasks to try to modify them.
+
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::::: discussion
+
+## Try this ...
+
+Are the concurrent tasks, spawned by the last code, running truly in parallel?
+
+The answer is: it depends on the number of cores available to your job. To verify this, let's modify the code
+to get the tasks into an infinite loop.
+
+```chpl
+begin
+{
+  var c=0;
+  while c>-1
+  {
+       c+=1;
+      //writeln('thread 1: ',x+c);
+   }
+}
+```
+
+Now submit your job asking for different amount of resources, and use system tools such as `top`or `ps` to
+monitor the execution of the code.
+
+:::::::::::::::::::::::::::::::::::::::::::::::::::
+
+::::::::::::::::::::::::::::::::::::: callout
+
+To maximise performance, start as many tasks as cores are available.
+
+::::::::::::::::::::::::::::::::::::::::::::::::
+
+A slightly more structured way to start concurrent tasks in Chapel is by using the `cobegin`statement. Here
+you can start a block of concurrent tasks, one for each statement inside the curly brackets. The main
+difference between the `begin`and `cobegin` statements is that with the `cobegin`, all the spawned tasks are
+synchronised at the end of the statement, i.e. the main thread won't continue its execution until all tasks
+are done.
+
+```chpl
+var x=0;
+writeln("This is the main thread, my value of x is ",x);
+
+cobegin
+{
+  {
+    var x=5;
+    writeln("this is task 1, my value of x is ",x);
+  }
+  writeln("this is task 2, my value of x is ",x);
+}
+
+writeln("this message won't appear until all tasks are done...");
+```
+
+```bash
+>> chpl cobegin_example.chpl -o cobegin_example
+>> ./cobegin_example
+```
+
+```output
+This is the main thread, my value of x is 0
+this is task 2, my value of x is 0
+this is task 1, my value of x is 5
+this message won't appear until all tasks are done...
+```
+
+As you may have conclude from the Discussion exercise above, the variables declared inside a task are
+accessible only by the task, while those variables declared in the main task are accessible to all tasks.
+
+The last, and most useful way to start concurrent/parallel tasks in Chapel, is the `coforall` loop. This is a
```

Comment:
Lesson introduces `coforall`, without having introduced `forall` at any earlier stage. Consider working that in.

Reply:
These lessons are broken into two parts: first task parallelism, and then data parallelism. The `forall` loop is introduced in 21-locales.md as part of the data-parallelism lesson. I think introducing it here will be a mistake, as it will make this chapter more complex and will duplicate the material. Now, many people suggested that data parallelism should be taught first as it is much easier than task parallelism. I tend to agree with this, and this is what I usually do when I don't have a full day to cover Chapel. Refactoring lessons with data parallelism first will be a major effort, and I think it's worth its own PR if we decide to do this in the future.